pkg/certgenerator/v1beta1/generator.go

```diff
@@ -0,0 +1,215 @@
+		certsReady:  certsReady,
+	})
+}
+
+// generate generates certificates for the admission webhooks.
+func (c *CertGenerator) generate(ctx context.Context) error {
+	controllerService := &corev1.Service{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.serviceName, Namespace: c.namespace}, controllerService); err != nil {
+		return fmt.Errorf("%w: %v", errServiceNotFound, err)
+	}
+
+	certExist, err := c.isCertExist(ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errCertCheckFail, err)
+	}
+	if !certExist {
+		c.fullServiceDomain = strings.Join([]string{c.serviceName, c.namespace, "svc"}, ".")
+
+		if err = c.createCert(); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertFail, err)
+		}
+		if err = c.updateCertSecret(ctx); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertSecretFail, err)
+		}
+	}
+	if err = c.injectCert(ctx); err != nil {
+		return fmt.Errorf("%w: %v", errInjectCertError, err)
+	}
+	return nil
+}
+
+// isCertExist checks if a secret embedded certs already exists.
+// For example, it will return true if the katib-controller is created with enabled leader-election
+// since another controller pod will create the secret.
+func (c *CertGenerator) isCertExist(ctx context.Context) (bool, error) {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.secretName, Namespace: c.namespace}, secret); err != nil {
+		return false, err
+	}
+	key := secret.Data[serverKeyName]
+	cert := secret.Data[serverCertName]
+	if len(key) != 0 && len(cert) != 0 {
+		c.certs = &certificates{
+			keyPem:  key,
+			certPem: cert,
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// createCert creates the self-signed certificate and private key.
+func (c *CertGenerator) createCert() error {
+	now := time.Now()
+	template := &x509.Certificate{
+		SerialNumber: big.NewInt(0),
+		Subject: pkix.Name{
+			CommonName: c.fullServiceDomain,
+		},
+		DNSNames: []string{
+			c.fullServiceDomain,
+		},
+		NotBefore:   now,
+		NotAfter:    now.Add(24 * time.Hour * 365 * 10),
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+	}
+
+	klog.Info("Generating self-signed public certificate and private key.")
+	rawKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+
+	der, err := x509.CreateCertificate(rand.Reader, template, template, rawKey.Public(), rawKey)
+	if err != nil {
+		return err
+	}
+	if c.certs, err = encode(rawKey, der); err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateCertSecret updates Secret embedded tls.key and tls.crt.
+func (c *CertGenerator) updateCertSecret(ctx context.Context) error {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.secretName, Namespace: c.namespace}, secret); err != nil {
+		return err
+	}
+	if len(secret.Data) == 0 {
+		secret.Data = make(map[string][]byte, 2)
+	}
```

Comment:
Do you need to verify that secret data is empty ?
I think, we can always override secret data with `tls.key` and `tls.crt`

Reply:
It makes sense.

Reply:
Done.